File: typescript/bc-crypto/src/hash.ts
import {
    createHash,
    createHmac,
    hkdfSync,
    pbkdf2Sync,
} from 'node:crypto';

import { type BytesLike, toBytes } from './bytes.js';

export const CRC32_SIZE = 4;
export const SHA256_SIZE = 32;
export const SHA512_SIZE = 64;

const CRC32_TABLE = new Uint32Array(256);
for (let i = 0; i < 256; i++) {
    let c = i;
    for (let j = 0; j < 8; j++) {
        c = (c & 1) !== 0 ? (0xedb88320 ^ (c >>> 1)) : (c >>> 1);
    }
    CRC32_TABLE[i] = c >>> 0;
}

/** Computes the CRC-32 checksum of the given data. */
export function crc32(data: BytesLike): number {
    const bytes = toBytes(data);
    let c = 0xffffffff;
    for (const byte of bytes) {
        c = CRC32_TABLE[(c ^ byte) & 0xff] ^ (c >>> 8);
    }
    return (c ^ 0xffffffff) >>> 0;
}

/** Computes the CRC-32 checksum bytes in big-endian or little-endian. */
export function crc32DataOpt(
    data: BytesLike,
    littleEndian: boolean,
): Uint8Array {
    const checksum = crc32(data);
    const out = new Uint8Array(CRC32_SIZE);
    const view = new DataView(out.buffer, out.byteOffset, out.byteLength);
    view.setUint32(0, checksum, littleEndian);
    return out;
}

/** Computes the CRC-32 checksum bytes in big-endian format. */
export function crc32Data(data: BytesLike): Uint8Array {
    return crc32DataOpt(data, false);
}

/** Computes the SHA-256 digest of the input buffer. */
export function sha256(data: BytesLike): Uint8Array {
    return new Uint8Array(createHash('sha256').update(toBytes(data)).digest());
}

/** Computes the double SHA-256 digest of the input buffer. */
export function doubleSha256(message: BytesLike): Uint8Array {
    return sha256(sha256(message));
}

/** Computes the SHA-512 digest of the input buffer. */
export function sha512(data: BytesLike): Uint8Array {
    return new Uint8Array(createHash('sha512').update(toBytes(data)).digest());
}

/** Computes the HMAC-SHA-256 for the given key and message. */
export function hmacSha256(key: BytesLike, message: BytesLike): Uint8Array {
    return new Uint8Array(
        createHmac('sha256', toBytes(key)).update(toBytes(message)).digest(),
    );
}

/** Computes the HMAC-SHA-512 for the given key and message. */
export function hmacSha512(key: BytesLike, message: BytesLike): Uint8Array {
    return new Uint8Array(
        createHmac('sha512', toBytes(key)).update(toBytes(message)).digest(),
    );
}

/** Computes the PBKDF2-HMAC-SHA-256 for the given password. */
export function pbkdf2HmacSha256(
    pass: BytesLike,
    salt: BytesLike,
    iterations: number,
    keyLen: number,
): Uint8Array {
    return new Uint8Array(
        pbkdf2Sync(toBytes(pass), toBytes(salt), iterations, keyLen, 'sha256'),
    );
}

/** Computes the PBKDF2-HMAC-SHA-512 for the given password. */
export function pbkdf2HmacSha512(
    pass: BytesLike,
    salt: BytesLike,
    iterations: number,
    keyLen: number,
): Uint8Array {
    return new Uint8Array(
        pbkdf2Sync(toBytes(pass), toBytes(salt), iterations, keyLen, 'sha512'),
    );
}

/** Computes the HKDF-HMAC-SHA-256 for the given key material. */
export function hkdfHmacSha256(
    keyMaterial: BytesLike,
    salt: BytesLike,
    keyLen: number,
): Uint8Array {
    return new Uint8Array(
        hkdfSync(
            'sha256',
            toBytes(keyMaterial),
            toBytes(salt),
            new Uint8Array(0),
            keyLen,
        ),
    );
}

/** Computes the HKDF-HMAC-SHA-512 for the given key material. */
export function hkdfHmacSha512(
    keyMaterial: BytesLike,
    salt: BytesLike,
    keyLen: number,
): Uint8Array {
    return new Uint8Array(
        hkdfSync(
            'sha512',
            toBytes(keyMaterial),
            toBytes(salt),
            new Uint8Array(0),
            keyLen,
        ),
    );
}

File: csharp/BCCrypto/BCCrypto/Hash.cs
using System.Buffers.Binary;
using System.IO.Hashing;
using System.Security.Cryptography;

namespace BlockchainCommons.BCCrypto;

/// <summary>
/// Cryptographic hash functions and key derivation utilities.
/// </summary>
public static class Hash
{
    public const int Crc32Size = 4;
    public const int Sha256Size = 32;
    public const int Sha512Size = 64;

    /// <summary>Computes the CRC-32 checksum of the given data.</summary>
    /// <param name="data">The data to hash.</param>
    /// <returns>The CRC-32 checksum.</returns>
    public static uint Crc32(ReadOnlySpan<byte> data)
    {
        return System.IO.Hashing.Crc32.HashToUInt32(data);
    }

    /// <summary>
    /// Computes the CRC-32 checksum of the given data, returning the hash as a
    /// 4-byte array in big-endian or little-endian format.
    /// </summary>
    /// <param name="data">The data to hash.</param>
    /// <param name="littleEndian">If <c>true</c>, return the checksum in little-endian byte order.</param>
    /// <returns>A 4-byte CRC-32 checksum.</returns>
    public static byte[] Crc32DataOpt(ReadOnlySpan<byte> data, bool littleEndian)
    {
        uint checksum = Crc32(data);
        byte[] result = new byte[Crc32Size];
        if (littleEndian)
            BinaryPrimitives.WriteUInt32LittleEndian(result, checksum);
        else
            BinaryPrimitives.WriteUInt32BigEndian(result, checksum);
        return result;
    }

    /// <summary>
    /// Computes the CRC-32 checksum of the given data, returning the hash as a
    /// 4-byte array in big-endian format.
    /// </summary>
    /// <param name="data">The data to hash.</param>
    /// <returns>A 4-byte CRC-32 checksum in big-endian byte order.</returns>
    public static byte[] Crc32Data(ReadOnlySpan<byte> data)
    {
        return Crc32DataOpt(data, false);
    }

    /// <summary>Computes the SHA-256 digest of the input data.</summary>
    /// <param name="data">The data to hash.</param>
    /// <returns>A 32-byte SHA-256 digest.</returns>
    public static byte[] Sha256(ReadOnlySpan<byte> data)
    {
        return SHA256.HashData(data);
    }

    /// <summary>Computes the double SHA-256 digest of the input data.</summary>
    /// <param name="message">The data to hash.</param>
    /// <returns>A 32-byte double-SHA-256 digest.</returns>
    public static byte[] DoubleSha256(ReadOnlySpan<byte> message)
    {
        return Sha256(Sha256(message));
    }

    /// <summary>Computes the SHA-512 digest of the input data.</summary>
    /// <param name="data">The data to hash.</param>
    /// <returns>A 64-byte SHA-512 digest.</returns>
    public static byte[] Sha512(ReadOnlySpan<byte> data)
    {
        return SHA512.HashData(data);
    }

    /// <summary>Computes the HMAC-SHA-256 for the given key and message.</summary>
    /// <param name="key">The HMAC key.</param>
    /// <param name="message">The message to authenticate.</param>
    /// <returns>A 32-byte HMAC-SHA-256 tag.</returns>
    public static byte[] HmacSha256(ReadOnlySpan<byte> key, ReadOnlySpan<byte> message)
    {
        return HMACSHA256.HashData(key, message);
    }

    /// <summary>Computes the HMAC-SHA-512 for the given key and message.</summary>
    /// <param name="key">The HMAC key.</param>
    /// <param name="message">The message to authenticate.</param>
    /// <returns>A 64-byte HMAC-SHA-512 tag.</returns>
    public static byte[] HmacSha512(ReadOnlySpan<byte> key, ReadOnlySpan<byte> message)
    {
        return HMACSHA512.HashData(key, message);
    }

    /// <summary>Computes the PBKDF2-HMAC-SHA-256 key derivation.</summary>
    /// <param name="pass">The password.</param>
    /// <param name="salt">The salt.</param>
    /// <param name="iterations">The number of PBKDF2 iterations.</param>
    /// <param name="keyLen">The desired output key length in bytes.</param>
    /// <returns>The derived key.</returns>
    public static byte[] Pbkdf2HmacSha256(
        ReadOnlySpan<byte> pass,
        ReadOnlySpan<byte> salt,
        uint iterations,
        int keyLen)
    {
        return Rfc2898DeriveBytes.Pbkdf2(pass, salt, (int)iterations, HashAlgorithmName.SHA256, keyLen);
    }

    /// <summary>Computes the PBKDF2-HMAC-SHA-512 key derivation.</summary>
    /// <param name="pass">The password.</param>
    /// <param name="salt">The salt.</param>
    /// <param name="iterations">The number of PBKDF2 iterations.</param>
    /// <param name="keyLen">The desired output key length in bytes.</param>
    /// <returns>The derived key.</returns>
    public static byte[] Pbkdf2HmacSha512(
        ReadOnlySpan<byte> pass,
        ReadOnlySpan<byte> salt,
        uint iterations,
        int keyLen)
    {
        return Rfc2898DeriveBytes.Pbkdf2(pass, salt, (int)iterations, HashAlgorithmName.SHA512, keyLen);
    }

    /// <summary>Computes the HKDF-HMAC-SHA-256 key derivation.</summary>
    /// <param name="keyMaterial">The input key material.</param>
    /// <param name="salt">The salt value.</param>
    /// <param name="keyLen">The desired output key length in bytes.</param>
    /// <returns>The derived key.</returns>
    public static byte[] HkdfHmacSha256(
        ReadOnlySpan<byte> keyMaterial,
        ReadOnlySpan<byte> salt,
        int keyLen)
    {
        byte[] output = new byte[keyLen];
        HKDF.DeriveKey(HashAlgorithmName.SHA256, keyMaterial, output, salt, ReadOnlySpan<byte>.Empty);
        return output;
    }

    /// <summary>Computes the HKDF-HMAC-SHA-512 key derivation.</summary>
    /// <param name="keyMaterial">The input key material.</param>
    /// <param name="salt">The salt value.</param>
    /// <param name="keyLen">The desired output key length in bytes.</param>
    /// <returns>The derived key.</returns>
    public static byte[] HkdfHmacSha512(
        ReadOnlySpan<byte> keyMaterial,
        ReadOnlySpan<byte> salt,
        int keyLen)
    {
        byte[] output = new byte[keyLen];
        HKDF.DeriveKey(HashAlgorithmName.SHA512, keyMaterial, output, salt, ReadOnlySpan<byte>.Empty);
        return output;
    }
}
